### PR TITLE
Bug 2000276: EncryptionStateControllerDegraded: failed to get converged static pod revision

### PR DESCRIPTION
**Pull request by p0lyn0mial**
It turns out that when a node dies or is disconnected from the rest of the cluster, Kubernetes (kubelet) applies a policy for setting the phase of all Pods on the lost node to Failed.
Pods in the Failed state cannot transition back to Running state.
In addition, failed Pods are note GC'ed. The API objects remain in the cluster's API until a human or controller process explicitly removes them.
Thus we need to tolerate pods in Succeeded or Failed phase only if their revision <= the running pods. Otherwise, we won't make any progress.

See the following for more:
 https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/
 https://bugzilla.redhat.com/show_bug.cgi?id=2000276

### CODE REVIEW

File: pkg/operator/encryption/deployer/revisionedpod_test.go [hunk 1]
```diff
@@ -132,7 +160,7 @@ func TestCategorizePods(t *testing.T) {
 				t.Errorf("categorizePods() gotProgressing = %v, want %v", gotProgressing, tt.wantCategorizeProgressing)
 			}
 
-			if err != nil {
+			if err == nil {
```
Comment (p0lyn0mial) [inline pkg/operator/encryption/deployer/revisionedpod_test.go:163]: I think this was wrong

File: pkg/operator/encryption/deployer/revisionedpod.go [hunk 2]
```diff
@@ -141,7 +141,13 @@ func (d *RevisionLabelPodDeployer) HasSynced() bool {
...
 //   2. All master nodes have a running pod
 //   3. There are no pending or unknown pods
-//   4. All succeeded and failed pods have revisions that are before the running pods
+//   4. We tolerate pods in Succeeded or Failed phase only if their revision <= the running pods
+//      It turns out that when a node dies or is disconnected from the rest of the cluster, Kubernetes (kubelet) applies a policy for setting the phase of all Pods on the lost node to Failed.
+//      Pods in the Failed state cannot transition back to Running state.
+//      In addition, failed Pods are note GC'ed. The API objects remain in the cluster's API until a human or controller process explicitly removes them.
+//      See the following for more:
+//        https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/
+//        https://bugzilla.redhat.com/show_bug.cgi?id=2000276
```
Comment (sttts) [inline pkg/operator/encryption/deployer/revisionedpod.go:150]: the BZ is about static pods. The pod life-cycle is about non-static pods, isn't it?

Comment (p0lyn0mial) [reply pkg/operator/encryption/deployer/revisionedpod.go:150]: the title is wrong then. The description points that there were `pods` in `InvalidNodeInfo` state in the `openshift-apiserver` ns

File: pkg/operator/encryption/deployer/revisionedpod_test.go (hunk 1 above)
Comment (dgrisonnet) [reply pkg/operator/encryption/deployer/revisionedpod_test.go:163]: I agree

File: pkg/operator/encryption/deployer/revisionedpod_test.go [hunk 3]
```diff
@@ -132,7 +160,7 @@ func TestCategorizePods(t *testing.T) {
...
 				rev, err := getAPIServerRevisionOfAllInstances("revision", tt.nodes, tt.pods)
 				if (err != nil) != tt.wantCategorizeErr {
```
Comment (dgrisonnet) [inline pkg/operator/encryption/deployer/revisionedpod_test.go:165]:
```suggestion
				if (err != nil) != wantGetAPIServerRevisionOfAllInstancesError {
```

**Review by dgrisonnet — COMMENTED** [review]
The fix makes sense to me.

As for the unit tests, I am not sure how much effort it would take, but it would be great to split the current `TestCategorizePods` into two different tests:
- TestCategorizePods
- TestGetAPIServerRevisionOfAllInstances

The current approach where a call to `getAPIServerRevisionOfAllInstances` was added alongside some additional checks goes against unit testing principles as we are testing two different things in one test. As we are seeing with this PR, it is confusing and prone to errors. I wouldn't block this PR with this refactor, but it would be great to at least have a follow-up PR improving that.

**Comment by dgrisonnet** [comment]
I will address the concerns I raised in a follow-up PR.

/lgtm

**Pull request merged**